{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "150ea3dd-bf90-4c56-a063-00c645a8097b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime\n",
    "import sys\n",
    "from pathlib import Path\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "ROOT_DIR = str(Path().resolve().parent)\n",
    "\n",
    "sys.path.append('../')\n",
    "plt.style.use(f'./project.mplstyle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f2eeb57b-8bc0-45f2-8166-19a18bb1c06f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<contextlib._GeneratorContextManager at 0x1037c6e20>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import importlib.resources\n",
    "importlib.resources.path(\"asd\", \"hmm/hmm_signal_peptide.joblib\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daf047c2-b946-4f4c-ac39-50b3a77db91f",
   "metadata": {},
   "source": [
    "# Genomes to Growth Conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cd34cb9f-596a-4b5b-95fe-02175d7ce797",
   "metadata": {},
   "outputs": [],
   "source": [
    "prefix = \"GCA_000172155.1_ASM17215v1\"\n",
    "contig_filepath = f\"../tests/test_data/{prefix}_genomic.fna.gz\"\n",
    "protein_filepath = f\"../tests/test_data/{prefix}_protein.faa.gz\"\n",
    "\n",
    "# prefix = \"GCA_016767175.1_ASM1676717v1\"\n",
    "# prefix =\"GCA_900096925.1_IMG-taxon_2616645037_annotated_assembly\"\n",
    "# contig_filepath = f\"../data/genomes/genbank/bacteria/{'_'.join(prefix.split('_')[0:2])}/{prefix}_genomic.fna.gz\"\n",
    "# protein_filepath = f\"../data/genomes/genbank/bacteria/{'_'.join(prefix.split('_')[0:2])}/{prefix}_protein.faa.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed667fc8-51cc-4ed1-8937-c5098c645307",
   "metadata": {},
   "source": [
    "## Predict Growth Conditions\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7debab9a-dd1a-41ad-bfc8-bc9c235efd0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-03-17 15:35:24 INFO Measuring features from:\n",
      "\t../data/genomes/genbank/bacteria/GCA_900096925.1/GCA_900096925.1_IMG-taxon_2616645037_annotated_assembly_genomic.fna.gz\n",
      "\t../data/genomes/genbank/bacteria/GCA_900096925.1/GCA_900096925.1_IMG-taxon_2616645037_annotated_assembly_protein.faa.gz\n",
      "2024-03-17 15:35:24 INFO GCA_900096925.1_IMG-taxon_2616645037_annotated_assembly_genomic.fna.gz: Identifying protein localization\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-03-17 15:35:24.699899\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-03-17 15:35:26 INFO GCA_900096925.1_IMG-taxon_2616645037_annotated_assembly_genomic.fna.gz: Collecting genome statistics\n",
      "2024-03-17 15:35:27 INFO GCA_900096925.1_IMG-taxon_2616645037_annotated_assembly_genomic.fna.gz: Collecting protein statistics\n",
      "2024-03-17 15:35:27 INFO Predicting growth conditions\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-03-17 15:35:27.137258\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value</th>\n",
       "      <th>error</th>\n",
       "      <th>is_novel</th>\n",
       "      <th>warning</th>\n",
       "      <th>units</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>oxygen</th>\n",
       "      <td>tolerant</td>\n",
       "      <td>0.831445</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>probability</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ph_max</th>\n",
       "      <td>9.339235</td>\n",
       "      <td>1.231827</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>pH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ph_min</th>\n",
       "      <td>5.854202</td>\n",
       "      <td>0.919082</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>pH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ph_optimum</th>\n",
       "      <td>7.608927</td>\n",
       "      <td>0.884935</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>pH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>salinity_max</th>\n",
       "      <td>7.855166</td>\n",
       "      <td>4.017368</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>% w/v NaCl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>salinity_min</th>\n",
       "      <td>2.206405</td>\n",
       "      <td>2.307733</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>% w/v NaCl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>salinity_optimum</th>\n",
       "      <td>3.10798</td>\n",
       "      <td>2.446439</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>% w/v NaCl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>temperature_max</th>\n",
       "      <td>45.18426</td>\n",
       "      <td>6.235459</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>temperature_min</th>\n",
       "      <td>18.253843</td>\n",
       "      <td>7.838025</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>temperature_optimum</th>\n",
       "      <td>34.09345</td>\n",
       "      <td>6.072984</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         value     error is_novel warning        units\n",
       "oxygen                tolerant  0.831445    False    None  probability\n",
       "ph_max                9.339235  1.231827    False    None           pH\n",
       "ph_min                5.854202  0.919082    False    None           pH\n",
       "ph_optimum            7.608927  0.884935    False    None           pH\n",
       "salinity_max          7.855166  4.017368    False    None   % w/v NaCl\n",
       "salinity_min          2.206405  2.307733    False    None   % w/v NaCl\n",
       "salinity_optimum       3.10798  2.446439    False    None   % w/v NaCl\n",
       "temperature_max       45.18426  6.235459    False    None            C\n",
       "temperature_min      18.253843  7.838025    False    None            C\n",
       "temperature_optimum   34.09345  6.072984    False    None            C"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from genome_spot.genome_spot import GenomeSPOT, run_genome_spot\n",
    "\n",
    "print(datetime.datetime.now())\n",
    "predictions, genome_features = run_genome_spot(\n",
    "        fna_path=contig_filepath,\n",
    "        faa_path=protein_filepath,\n",
    "        features_json=None,\n",
    "        path_to_models=f\"../models\",\n",
    "        #save_intermediate=False,\n",
    "    )\n",
    "print(datetime.datetime.now())\n",
    "\n",
    "pd.DataFrame(predictions).T.sort_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2a9514d-3c63-4cd8-a51c-2ff0a7a717da",
   "metadata": {},
   "source": [
    "## Measure Genome Properties"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a138d73a-0420-47c6-8c18-9ccfdfa21e0b",
   "metadata": {},
   "source": [
    "### 1 Protein Sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a4aab02c-c9a7-45d6-9720-df32493e1930",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'pi': 5.382106971740723,\n",
       " 'zc': 0.19821826280623608,\n",
       " 'nh2o': 0.07137639198218272,\n",
       " 'gravy': -0.6514476614699332,\n",
       " 'thermostable_freq': 0.356347438752784,\n",
       " 'length': 449,\n",
       " 'is_exported': False,\n",
       " 'aa_M': 0.0200445434298441,\n",
       " 'aa_A': 0.08463251670378619,\n",
       " 'aa_Q': 0.10690423162583519,\n",
       " 'aa_D': 0.0757238307349666,\n",
       " 'aa_V': 0.0935412026726058,\n",
       " 'aa_K': 0.0334075723830735,\n",
       " 'aa_T': 0.0400890868596882,\n",
       " 'aa_P': 0.0467706013363029,\n",
       " 'aa_Y': 0.026726057906458798,\n",
       " 'aa_R': 0.062360801781737196,\n",
       " 'aa_I': 0.031180400890868598,\n",
       " 'aa_L': 0.08240534521158129,\n",
       " 'aa_E': 0.05790645879732739,\n",
       " 'aa_W': 0.0022271714922048997,\n",
       " 'aa_G': 0.08685968819599109,\n",
       " 'aa_N': 0.035634743875278395,\n",
       " 'aa_F': 0.011135857461024499,\n",
       " 'aa_H': 0.035634743875278395,\n",
       " 'aa_S': 0.066815144766147}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from genome_spot.bioinformatics.protein import Protein\n",
    "\n",
    "sequence = ''.join(\"\"\"MAAQDVKQQTPYRVIQLEWDAEKGERNEAVGNFDELVTHHPKSNSDAHLVDGKVVGGQAG\n",
    "RTLGVVGGEIQEIEVSKAGKDYGLRPDQVLLKKDFMLEDSRLPSGPSSRSLDVPSPVAGV\n",
    "VGTVNTSKGLVDVLDREGGDVILRVRHMSPLHVKAGDQVEYGQALGVQGKQATGAIHVHM\n",
    "EVDSRYYQHYENYVGDLVSGRLSIDAERRDRGIEPRPFVDDGTIRIGGSSEMVQKVQQTL\n",
    "NAEGYRGADNQPLQEDGVYRLSMQAAVINYQQAHGLSQTGDIDPATLQQIAPRTFPPELN\n",
    "REDHNATPTYRNLQGAVPSQDPLHRQAEEDVRRLEQSLGRDYDDNSARLAASSAHLAKAN\n",
    "GLTQIDHVVLSNQTAAVGKGENVFVVQGALDNPAHLMAHMKTSDAIAQPVEQSLSQLQTL\n",
    "SETQRQQQAQQQSQQQDQQQLSAPQHRMV\"\"\".split('\\n'))\n",
    "\n",
    "protein = Protein(sequence)\n",
    "protein.protein_metrics()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c297cdc7-d2d9-4fec-94c5-e86987a7208c",
   "metadata": {},
   "source": [
    "### 1 Signal Peptide Prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ba454b1f-0731-46ef-9306-09e6089605ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(True, 'MNKTLIAAAVAGIVLLASNAQA')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from genome_spot.bioinformatics.signal_peptide import SignalPeptideHMM\n",
    "\n",
    "partial_sequence = 'MNKTLIAAAVAGIVLLASNAQAQTVPEGYQLQQVLMMSRHNLRAPLANNG'\n",
    "\n",
    "signal_peptide_model = SignalPeptideHMM()\n",
    "is_exported, signal_end_index = signal_peptide_model.predict_signal_peptide(partial_sequence)\n",
    "signal_peptide = partial_sequence[:signal_end_index+1]\n",
    "is_exported, signal_peptide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fc15df8a-38e9-4ee1-b9bc-e9f160160209",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-130.73482767354764\n",
      "M B\n",
      "N N3\n",
      "K N2\n",
      "T N1\n",
      "L H\n",
      "I H\n",
      "A H\n",
      "A H\n",
      "A H\n",
      "V H\n",
      "A H\n",
      "G H\n",
      "I H\n",
      "V H\n",
      "L H\n",
      "L H\n",
      "A H\n",
      "S C\n",
      "N C\n",
      "A C3\n",
      "Q C2\n",
      "A C1\n",
      "Q M\n",
      "T M\n",
      "V M\n",
      "P M\n",
      "E M\n",
      "G M\n",
      "Y M\n",
      "Q M\n",
      "L M\n",
      "Q M\n",
      "Q M\n",
      "V M\n",
      "L M\n",
      "M M\n",
      "M M\n",
      "S M\n",
      "R M\n",
      "H M\n",
      "N M\n",
      "L M\n",
      "R M\n",
      "A M\n",
      "P M\n",
      "L M\n",
      "A M\n",
      "N M\n",
      "N M\n",
      "G M\n"
     ]
    }
   ],
   "source": [
    "formatted_sequence = signal_peptide_model._format_protein_sequence(partial_sequence)\n",
    "pred_states, log_prob = signal_peptide_model._predict_hidden_states(formatted_sequence)\n",
    "print(log_prob)\n",
    "for aa, state in zip(partial_sequence[:len(formatted_sequence)], pred_states):\n",
    "    print(aa,state)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7722bc5b-7b01-4098-aa68-eddaa1305096",
   "metadata": {},
   "source": [
    "### 1 DNA Sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c346e51b-acb9-438d-a5b5-0a7eda11dd3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'nt_length': 585,\n",
       " 'pur_pyr_transition_freq': 0.4897260273972603,\n",
       " 'nt_C': 0.558974358974359,\n",
       " 'nt_A': 0.441025641025641}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from genome_spot.bioinformatics.dna import DNA\n",
    "\n",
    "sequence = \"GGATGGACGGAGGAATTCCTCAAGGAAGTCGGGCCCGCGCTGGTGGTACTCGGTCCAGGCTTCTTGCACGAAGAAGTCTCCGACCGCGCCTCTCTCCACCCTCCTGGCAAAATCGGCCAGTGACTTGATGCCGATGTGGTAGATGAAGCCGGTGTCGAGTACGCCCTTGGCGAAGGTCGGGTCCGCCAGCCTCTCGGGAAAATTTTCTTCGATGTTCAGATAGTAGCCGCGCATGCTGACGCCACGGCTATCTAAAGTGAACATTCCGTCCTTCAGAATGCAATCGGCATGGTGCCAAAATCCAGAGCCCACCGCGCTGGCAATCTTGTTGTCGTCAACTGCGTCGGATCGGACGAGGCACTTCGTGAACGGCGCCGGCAATTCGGACCGTCGAGTGTAATTACGGATTGCGGACCCAGACGGGCGCGTGGCGTGGCCGTTATGCCCAGCATTGACGGAGAACAGGTGGATTGCCCCCACTCCACCCGGAAAGCGTGAGATAAACTCTCGGATCGTGTCGTCGTTGCGCAGTAAAATAAATTCGTCAACATCGAGAAATATGTAATTACGGCTCAGATGTTTAAA\"\n",
    "dna = DNA(sequence)\n",
    "dna.nucleotide_metrics()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8a681fad-37f5-4b19-92ff-2811085f0bd9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'nt_length': 13,\n",
       " 'pur_pyr_transition_freq': 0.5833333333333334,\n",
       " 'nt_A': 0.46153846153846156,\n",
       " 'nt_C': 0.5384615384615384}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DNA(\"ACTAGCGACTAGC\").nucleotide_metrics()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d08ab94f-ff2a-4ffd-af08-7f459e0e2a20",
   "metadata": {},
   "source": [
    "### All Protein and DNA Sequences\n",
    "\n",
    "Note: It is expected and good to see NaN for many features because they cannot be computed by cellular localization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2703afe3-c023-42db-b56b-241a9a2a4fcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-03-15 15:37:22 INFO GCA_000172155.1_ASM17215v1_genomic.fna.gz: Identifying protein localization\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-03-15 15:37:22.581463\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-03-15 15:37:25 INFO GCA_000172155.1_ASM17215v1_genomic.fna.gz: Collecting genome statistics\n",
      "2024-03-15 15:37:26 INFO GCA_000172155.1_ASM17215v1_genomic.fna.gz: Collecting protein statistics\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-03-15 15:37:27.009166\n"
     ]
    }
   ],
   "source": [
    "from genome_spot.bioinformatics.genome import Genome\n",
    "from genome_spot.bioinformatics.signal_peptide import SignalPeptideHMM\n",
    "\n",
    "genome_calc = Genome(\n",
    "    contig_filepath=contig_filepath, \n",
    "    protein_filepath=protein_filepath,\n",
    ")\n",
    "\n",
    "print(datetime.datetime.now())\n",
    "genome_features = genome_calc.measure_genome_features()\n",
    "print(datetime.datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5f7b26a8-ff8e-4572-9569-cef2ff308f84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>all</th>\n",
       "      <th>extracellular_soluble</th>\n",
       "      <th>intracellular_soluble</th>\n",
       "      <th>membrane</th>\n",
       "      <th>diff_extra_intra</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>nt_length</th>\n",
       "      <td>8.220859e+06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pur_pyr_transition_freq</th>\n",
       "      <td>4.796972e-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nt_C</th>\n",
       "      <td>6.031152e-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nt_A</th>\n",
       "      <td>3.968848e-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total_proteins</th>\n",
       "      <td>6.519000e+03</td>\n",
       "      <td>635.000000</td>\n",
       "      <td>5.336000e+03</td>\n",
       "      <td>548.000000</td>\n",
       "      <td>-4.701000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total_protein_length</th>\n",
       "      <td>2.366909e+06</td>\n",
       "      <td>285520.000000</td>\n",
       "      <td>1.921398e+06</td>\n",
       "      <td>159991.000000</td>\n",
       "      <td>-1.635878e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pis_acidic</th>\n",
       "      <td>1.988035e-01</td>\n",
       "      <td>0.215748</td>\n",
       "      <td>2.113943e-01</td>\n",
       "      <td>0.056569</td>\n",
       "      <td>4.353729e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pis_neutral</th>\n",
       "      <td>4.914864e-01</td>\n",
       "      <td>0.615748</td>\n",
       "      <td>5.018741e-01</td>\n",
       "      <td>0.246350</td>\n",
       "      <td>1.138740e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pis_basic</th>\n",
       "      <td>3.097101e-01</td>\n",
       "      <td>0.168504</td>\n",
       "      <td>2.867316e-01</td>\n",
       "      <td>0.697080</td>\n",
       "      <td>-1.182277e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pis_3_4</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pis_4_5</th>\n",
       "      <td>9.633379e-02</td>\n",
       "      <td>0.114961</td>\n",
       "      <td>1.000750e-01</td>\n",
       "      <td>0.038321</td>\n",
       "      <td>1.488567e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pis_5_6</th>\n",
       "      <td>2.414481e-01</td>\n",
       "      <td>0.236220</td>\n",
       "      <td>2.610570e-01</td>\n",
       "      <td>0.056569</td>\n",
       "      <td>-2.483650e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pis_6_7</th>\n",
       "      <td>2.202792e-01</td>\n",
       "      <td>0.286614</td>\n",
       "      <td>2.254498e-01</td>\n",
       "      <td>0.093066</td>\n",
       "      <td>6.116440e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pis_7_8</th>\n",
       "      <td>8.712993e-02</td>\n",
       "      <td>0.133858</td>\n",
       "      <td>8.264618e-02</td>\n",
       "      <td>0.076642</td>\n",
       "      <td>5.121209e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pis_8_9</th>\n",
       "      <td>1.122872e-01</td>\n",
       "      <td>0.129134</td>\n",
       "      <td>1.083208e-01</td>\n",
       "      <td>0.131387</td>\n",
       "      <td>2.081302e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pis_9_10</th>\n",
       "      <td>1.774812e-01</td>\n",
       "      <td>0.085039</td>\n",
       "      <td>1.604198e-01</td>\n",
       "      <td>0.450730</td>\n",
       "      <td>-7.538042e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pis_10_11</th>\n",
       "      <td>4.509894e-02</td>\n",
       "      <td>0.011024</td>\n",
       "      <td>4.085457e-02</td>\n",
       "      <td>0.125912</td>\n",
       "      <td>-2.983095e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pis_11_12</th>\n",
       "      <td>1.994171e-02</td>\n",
       "      <td>0.003150</td>\n",
       "      <td>2.117691e-02</td>\n",
       "      <td>0.027372</td>\n",
       "      <td>-1.802731e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean_pi</th>\n",
       "      <td>7.200740e+00</td>\n",
       "      <td>6.742791</td>\n",
       "      <td>7.098613e+00</td>\n",
       "      <td>8.725832</td>\n",
       "      <td>-3.558220e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean_gravy</th>\n",
       "      <td>-1.747939e-01</td>\n",
       "      <td>-0.347193</td>\n",
       "      <td>-2.403479e-01</td>\n",
       "      <td>0.663289</td>\n",
       "      <td>-1.068450e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean_zc</th>\n",
       "      <td>-3.247701e-01</td>\n",
       "      <td>-0.092339</td>\n",
       "      <td>-2.673857e-01</td>\n",
       "      <td>-1.152867</td>\n",
       "      <td>1.750466e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean_nh2o</th>\n",
       "      <td>6.249488e-03</td>\n",
       "      <td>0.013570</td>\n",
       "      <td>7.763005e-03</td>\n",
       "      <td>-0.016970</td>\n",
       "      <td>5.806798e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean_protein_length</th>\n",
       "      <td>3.630785e+02</td>\n",
       "      <td>449.637795</td>\n",
       "      <td>3.600821e+02</td>\n",
       "      <td>291.954380</td>\n",
       "      <td>8.955571e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean_thermostable_freq</th>\n",
       "      <td>3.810484e-01</td>\n",
       "      <td>0.351282</td>\n",
       "      <td>3.825798e-01</td>\n",
       "      <td>0.415780</td>\n",
       "      <td>-3.129790e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>proportion_R_RK</th>\n",
       "      <td>5.884605e-01</td>\n",
       "      <td>0.483558</td>\n",
       "      <td>6.024070e-01</td>\n",
       "      <td>0.633337</td>\n",
       "      <td>-1.188490e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa_M</th>\n",
       "      <td>2.434063e-02</td>\n",
       "      <td>0.017372</td>\n",
       "      <td>2.201470e-02</td>\n",
       "      <td>0.032755</td>\n",
       "      <td>-4.643013e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa_S</th>\n",
       "      <td>6.170794e-02</td>\n",
       "      <td>0.060668</td>\n",
       "      <td>6.148287e-02</td>\n",
       "      <td>0.059413</td>\n",
       "      <td>-8.151978e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa_L</th>\n",
       "      <td>1.017539e-01</td>\n",
       "      <td>0.082423</td>\n",
       "      <td>1.016326e-01</td>\n",
       "      <td>0.140612</td>\n",
       "      <td>-1.921010e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa_Y</th>\n",
       "      <td>2.404187e-02</td>\n",
       "      <td>0.029761</td>\n",
       "      <td>2.382340e-02</td>\n",
       "      <td>0.022378</td>\n",
       "      <td>5.937607e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa_T</th>\n",
       "      <td>5.536257e-02</td>\n",
       "      <td>0.064883</td>\n",
       "      <td>5.509440e-02</td>\n",
       "      <td>0.050377</td>\n",
       "      <td>9.788192e-03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  all  extracellular_soluble  \\\n",
       "nt_length                8.220859e+06                    NaN   \n",
       "pur_pyr_transition_freq  4.796972e-01                    NaN   \n",
       "nt_C                     6.031152e-01                    NaN   \n",
       "nt_A                     3.968848e-01                    NaN   \n",
       "total_proteins           6.519000e+03             635.000000   \n",
       "total_protein_length     2.366909e+06          285520.000000   \n",
       "pis_acidic               1.988035e-01               0.215748   \n",
       "pis_neutral              4.914864e-01               0.615748   \n",
       "pis_basic                3.097101e-01               0.168504   \n",
       "pis_3_4                  0.000000e+00               0.000000   \n",
       "pis_4_5                  9.633379e-02               0.114961   \n",
       "pis_5_6                  2.414481e-01               0.236220   \n",
       "pis_6_7                  2.202792e-01               0.286614   \n",
       "pis_7_8                  8.712993e-02               0.133858   \n",
       "pis_8_9                  1.122872e-01               0.129134   \n",
       "pis_9_10                 1.774812e-01               0.085039   \n",
       "pis_10_11                4.509894e-02               0.011024   \n",
       "pis_11_12                1.994171e-02               0.003150   \n",
       "mean_pi                  7.200740e+00               6.742791   \n",
       "mean_gravy              -1.747939e-01              -0.347193   \n",
       "mean_zc                 -3.247701e-01              -0.092339   \n",
       "mean_nh2o                6.249488e-03               0.013570   \n",
       "mean_protein_length      3.630785e+02             449.637795   \n",
       "mean_thermostable_freq   3.810484e-01               0.351282   \n",
       "proportion_R_RK          5.884605e-01               0.483558   \n",
       "aa_M                     2.434063e-02               0.017372   \n",
       "aa_S                     6.170794e-02               0.060668   \n",
       "aa_L                     1.017539e-01               0.082423   \n",
       "aa_Y                     2.404187e-02               0.029761   \n",
       "aa_T                     5.536257e-02               0.064883   \n",
       "\n",
       "                         intracellular_soluble       membrane  \\\n",
       "nt_length                                  NaN            NaN   \n",
       "pur_pyr_transition_freq                    NaN            NaN   \n",
       "nt_C                                       NaN            NaN   \n",
       "nt_A                                       NaN            NaN   \n",
       "total_proteins                    5.336000e+03     548.000000   \n",
       "total_protein_length              1.921398e+06  159991.000000   \n",
       "pis_acidic                        2.113943e-01       0.056569   \n",
       "pis_neutral                       5.018741e-01       0.246350   \n",
       "pis_basic                         2.867316e-01       0.697080   \n",
       "pis_3_4                           0.000000e+00       0.000000   \n",
       "pis_4_5                           1.000750e-01       0.038321   \n",
       "pis_5_6                           2.610570e-01       0.056569   \n",
       "pis_6_7                           2.254498e-01       0.093066   \n",
       "pis_7_8                           8.264618e-02       0.076642   \n",
       "pis_8_9                           1.083208e-01       0.131387   \n",
       "pis_9_10                          1.604198e-01       0.450730   \n",
       "pis_10_11                         4.085457e-02       0.125912   \n",
       "pis_11_12                         2.117691e-02       0.027372   \n",
       "mean_pi                           7.098613e+00       8.725832   \n",
       "mean_gravy                       -2.403479e-01       0.663289   \n",
       "mean_zc                          -2.673857e-01      -1.152867   \n",
       "mean_nh2o                         7.763005e-03      -0.016970   \n",
       "mean_protein_length               3.600821e+02     291.954380   \n",
       "mean_thermostable_freq            3.825798e-01       0.415780   \n",
       "proportion_R_RK                   6.024070e-01       0.633337   \n",
       "aa_M                              2.201470e-02       0.032755   \n",
       "aa_S                              6.148287e-02       0.059413   \n",
       "aa_L                              1.016326e-01       0.140612   \n",
       "aa_Y                              2.382340e-02       0.022378   \n",
       "aa_T                              5.509440e-02       0.050377   \n",
       "\n",
       "                         diff_extra_intra  \n",
       "nt_length                             NaN  \n",
       "pur_pyr_transition_freq               NaN  \n",
       "nt_C                                  NaN  \n",
       "nt_A                                  NaN  \n",
       "total_proteins              -4.701000e+03  \n",
       "total_protein_length        -1.635878e+06  \n",
       "pis_acidic                   4.353729e-03  \n",
       "pis_neutral                  1.138740e-01  \n",
       "pis_basic                   -1.182277e-01  \n",
       "pis_3_4                      0.000000e+00  \n",
       "pis_4_5                      1.488567e-02  \n",
       "pis_5_6                     -2.483650e-02  \n",
       "pis_6_7                      6.116440e-02  \n",
       "pis_7_8                      5.121209e-02  \n",
       "pis_8_9                      2.081302e-02  \n",
       "pis_9_10                    -7.538042e-02  \n",
       "pis_10_11                   -2.983095e-02  \n",
       "pis_11_12                   -1.802731e-02  \n",
       "mean_pi                     -3.558220e-01  \n",
       "mean_gravy                  -1.068450e-01  \n",
       "mean_zc                      1.750466e-01  \n",
       "mean_nh2o                    5.806798e-03  \n",
       "mean_protein_length          8.955571e+01  \n",
       "mean_thermostable_freq      -3.129790e-02  \n",
       "proportion_R_RK             -1.188490e-01  \n",
       "aa_M                        -4.643013e-03  \n",
       "aa_S                        -8.151978e-04  \n",
       "aa_L                        -1.921010e-02  \n",
       "aa_Y                         5.937607e-03  \n",
       "aa_T                         9.788192e-03  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genome = pd.DataFrame(genome_features)\n",
    "df_genome.head(30)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5757327c-03af-40fd-a268-b7f5f01d5b3d",
   "metadata": {},
   "source": [
    "### All Proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7ea53949-6736-4741-aaea-3fb225fd70f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pi</th>\n",
       "      <th>zc</th>\n",
       "      <th>nh2o</th>\n",
       "      <th>gravy</th>\n",
       "      <th>thermostable_freq</th>\n",
       "      <th>length</th>\n",
       "      <th>is_exported</th>\n",
       "      <th>aa_M</th>\n",
       "      <th>aa_S</th>\n",
       "      <th>aa_L</th>\n",
       "      <th>...</th>\n",
       "      <th>aa_C</th>\n",
       "      <th>aa_G</th>\n",
       "      <th>aa_A</th>\n",
       "      <th>aa_H</th>\n",
       "      <th>aa_E</th>\n",
       "      <th>aa_R</th>\n",
       "      <th>aa_Q</th>\n",
       "      <th>aa_N</th>\n",
       "      <th>aa_F</th>\n",
       "      <th>aa_W</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABIZ01000001.1_1</th>\n",
       "      <td>7.551992</td>\n",
       "      <td>-0.179457</td>\n",
       "      <td>0.014524</td>\n",
       "      <td>-0.33754</td>\n",
       "      <td>0.386581</td>\n",
       "      <td>626</td>\n",
       "      <td>False</td>\n",
       "      <td>0.025559</td>\n",
       "      <td>0.059105</td>\n",
       "      <td>0.089457</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015974</td>\n",
       "      <td>0.08147</td>\n",
       "      <td>0.087859</td>\n",
       "      <td>0.017572</td>\n",
       "      <td>0.070288</td>\n",
       "      <td>0.063898</td>\n",
       "      <td>0.044728</td>\n",
       "      <td>0.031949</td>\n",
       "      <td>0.031949</td>\n",
       "      <td>0.01278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABIZ01000001.1_2</th>\n",
       "      <td>11.364825</td>\n",
       "      <td>-0.744643</td>\n",
       "      <td>0.070964</td>\n",
       "      <td>0.251786</td>\n",
       "      <td>0.464286</td>\n",
       "      <td>56</td>\n",
       "      <td>False</td>\n",
       "      <td>0.017857</td>\n",
       "      <td>0.089286</td>\n",
       "      <td>0.125</td>\n",
       "      <td>...</td>\n",
       "      <td>0.017857</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.089286</td>\n",
       "      <td>0.017857</td>\n",
       "      <td>0.017857</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.017857</td>\n",
       "      <td>0.035714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABIZ01000001.1_3</th>\n",
       "      <td>6.49581</td>\n",
       "      <td>-0.276424</td>\n",
       "      <td>-0.020354</td>\n",
       "      <td>-0.144702</td>\n",
       "      <td>0.39404</td>\n",
       "      <td>302</td>\n",
       "      <td>False</td>\n",
       "      <td>0.023179</td>\n",
       "      <td>0.046358</td>\n",
       "      <td>0.112583</td>\n",
       "      <td>...</td>\n",
       "      <td>0.016556</td>\n",
       "      <td>0.092715</td>\n",
       "      <td>0.096026</td>\n",
       "      <td>0.043046</td>\n",
       "      <td>0.062914</td>\n",
       "      <td>0.05298</td>\n",
       "      <td>0.019868</td>\n",
       "      <td>0.039735</td>\n",
       "      <td>0.029801</td>\n",
       "      <td>0.013245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABIZ01000001.1_4</th>\n",
       "      <td>6.166826</td>\n",
       "      <td>-0.120446</td>\n",
       "      <td>-0.054786</td>\n",
       "      <td>-0.380501</td>\n",
       "      <td>0.395543</td>\n",
       "      <td>359</td>\n",
       "      <td>False</td>\n",
       "      <td>0.022284</td>\n",
       "      <td>0.061281</td>\n",
       "      <td>0.094708</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011142</td>\n",
       "      <td>0.075209</td>\n",
       "      <td>0.075209</td>\n",
       "      <td>0.038997</td>\n",
       "      <td>0.066852</td>\n",
       "      <td>0.064067</td>\n",
       "      <td>0.030641</td>\n",
       "      <td>0.050139</td>\n",
       "      <td>0.038997</td>\n",
       "      <td>0.011142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABIZ01000001.1_5</th>\n",
       "      <td>5.399329</td>\n",
       "      <td>-0.281062</td>\n",
       "      <td>0.052735</td>\n",
       "      <td>-0.257522</td>\n",
       "      <td>0.411504</td>\n",
       "      <td>226</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00885</td>\n",
       "      <td>0.070796</td>\n",
       "      <td>0.132743</td>\n",
       "      <td>...</td>\n",
       "      <td>0.022124</td>\n",
       "      <td>0.070796</td>\n",
       "      <td>0.110619</td>\n",
       "      <td>0.00885</td>\n",
       "      <td>0.070796</td>\n",
       "      <td>0.084071</td>\n",
       "      <td>0.053097</td>\n",
       "      <td>0.017699</td>\n",
       "      <td>0.030973</td>\n",
       "      <td>0.022124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABIZ01000001.1_6515</th>\n",
       "      <td>9.24749</td>\n",
       "      <td>-0.947007</td>\n",
       "      <td>-0.111161</td>\n",
       "      <td>0.443796</td>\n",
       "      <td>0.394161</td>\n",
       "      <td>137</td>\n",
       "      <td>False</td>\n",
       "      <td>0.043796</td>\n",
       "      <td>0.051095</td>\n",
       "      <td>0.109489</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007299</td>\n",
       "      <td>0.072993</td>\n",
       "      <td>0.153285</td>\n",
       "      <td>0.014599</td>\n",
       "      <td>0.021898</td>\n",
       "      <td>0.065693</td>\n",
       "      <td>0.029197</td>\n",
       "      <td>0.007299</td>\n",
       "      <td>0.058394</td>\n",
       "      <td>0.029197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABIZ01000001.1_6516</th>\n",
       "      <td>8.159264</td>\n",
       "      <td>0.033143</td>\n",
       "      <td>0.083814</td>\n",
       "      <td>-0.695714</td>\n",
       "      <td>0.414286</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "      <td>0.014286</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>0.114286</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>0.042857</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABIZ01000001.1_6517</th>\n",
       "      <td>5.395009</td>\n",
       "      <td>-0.4275</td>\n",
       "      <td>0.021831</td>\n",
       "      <td>-0.223529</td>\n",
       "      <td>0.433824</td>\n",
       "      <td>136</td>\n",
       "      <td>False</td>\n",
       "      <td>0.036765</td>\n",
       "      <td>0.051471</td>\n",
       "      <td>0.132353</td>\n",
       "      <td>...</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>0.095588</td>\n",
       "      <td>0.066176</td>\n",
       "      <td>0.014706</td>\n",
       "      <td>0.080882</td>\n",
       "      <td>0.095588</td>\n",
       "      <td>0.051471</td>\n",
       "      <td>0.014706</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>0.022059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABIZ01000001.1_6518</th>\n",
       "      <td>4.872716</td>\n",
       "      <td>-0.37697</td>\n",
       "      <td>-0.011582</td>\n",
       "      <td>-0.013636</td>\n",
       "      <td>0.4</td>\n",
       "      <td>330</td>\n",
       "      <td>False</td>\n",
       "      <td>0.045455</td>\n",
       "      <td>0.060606</td>\n",
       "      <td>0.075758</td>\n",
       "      <td>...</td>\n",
       "      <td>0.024242</td>\n",
       "      <td>0.078788</td>\n",
       "      <td>0.087879</td>\n",
       "      <td>0.021212</td>\n",
       "      <td>0.075758</td>\n",
       "      <td>0.054545</td>\n",
       "      <td>0.018182</td>\n",
       "      <td>0.036364</td>\n",
       "      <td>0.036364</td>\n",
       "      <td>0.015152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABIZ01000001.1_6519</th>\n",
       "      <td>5.83051</td>\n",
       "      <td>0.141061</td>\n",
       "      <td>-0.092237</td>\n",
       "      <td>-0.414246</td>\n",
       "      <td>0.332402</td>\n",
       "      <td>358</td>\n",
       "      <td>False</td>\n",
       "      <td>0.03352</td>\n",
       "      <td>0.041899</td>\n",
       "      <td>0.061453</td>\n",
       "      <td>...</td>\n",
       "      <td>0.01676</td>\n",
       "      <td>0.097765</td>\n",
       "      <td>0.128492</td>\n",
       "      <td>0.036313</td>\n",
       "      <td>0.078212</td>\n",
       "      <td>0.039106</td>\n",
       "      <td>0.036313</td>\n",
       "      <td>0.039106</td>\n",
       "      <td>0.041899</td>\n",
       "      <td>0.011173</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6519 rows Ã— 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            pi        zc      nh2o     gravy  \\\n",
       "ABIZ01000001.1_1      7.551992 -0.179457  0.014524  -0.33754   \n",
       "ABIZ01000001.1_2     11.364825 -0.744643  0.070964  0.251786   \n",
       "ABIZ01000001.1_3       6.49581 -0.276424 -0.020354 -0.144702   \n",
       "ABIZ01000001.1_4      6.166826 -0.120446 -0.054786 -0.380501   \n",
       "ABIZ01000001.1_5      5.399329 -0.281062  0.052735 -0.257522   \n",
       "...                        ...       ...       ...       ...   \n",
       "ABIZ01000001.1_6515    9.24749 -0.947007 -0.111161  0.443796   \n",
       "ABIZ01000001.1_6516   8.159264  0.033143  0.083814 -0.695714   \n",
       "ABIZ01000001.1_6517   5.395009   -0.4275  0.021831 -0.223529   \n",
       "ABIZ01000001.1_6518   4.872716  -0.37697 -0.011582 -0.013636   \n",
       "ABIZ01000001.1_6519    5.83051  0.141061 -0.092237 -0.414246   \n",
       "\n",
       "                    thermostable_freq length is_exported      aa_M      aa_S  \\\n",
       "ABIZ01000001.1_1             0.386581    626       False  0.025559  0.059105   \n",
       "ABIZ01000001.1_2             0.464286     56       False  0.017857  0.089286   \n",
       "ABIZ01000001.1_3              0.39404    302       False  0.023179  0.046358   \n",
       "ABIZ01000001.1_4             0.395543    359       False  0.022284  0.061281   \n",
       "ABIZ01000001.1_5             0.411504    226       False   0.00885  0.070796   \n",
       "...                               ...    ...         ...       ...       ...   \n",
       "ABIZ01000001.1_6515          0.394161    137       False  0.043796  0.051095   \n",
       "ABIZ01000001.1_6516          0.414286     70       False  0.014286  0.057143   \n",
       "ABIZ01000001.1_6517          0.433824    136       False  0.036765  0.051471   \n",
       "ABIZ01000001.1_6518               0.4    330       False  0.045455  0.060606   \n",
       "ABIZ01000001.1_6519          0.332402    358       False   0.03352  0.041899   \n",
       "\n",
       "                         aa_L  ...      aa_C      aa_G      aa_A      aa_H  \\\n",
       "ABIZ01000001.1_1     0.089457  ...  0.015974   0.08147  0.087859  0.017572   \n",
       "ABIZ01000001.1_2        0.125  ...  0.017857  0.107143  0.089286  0.017857   \n",
       "ABIZ01000001.1_3     0.112583  ...  0.016556  0.092715  0.096026  0.043046   \n",
       "ABIZ01000001.1_4     0.094708  ...  0.011142  0.075209  0.075209  0.038997   \n",
       "ABIZ01000001.1_5     0.132743  ...  0.022124  0.070796  0.110619   0.00885   \n",
       "...                       ...  ...       ...       ...       ...       ...   \n",
       "ABIZ01000001.1_6515  0.109489  ...  0.007299  0.072993  0.153285  0.014599   \n",
       "ABIZ01000001.1_6516  0.114286  ...       NaN  0.057143  0.057143  0.042857   \n",
       "ABIZ01000001.1_6517  0.132353  ...  0.029412  0.095588  0.066176  0.014706   \n",
       "ABIZ01000001.1_6518  0.075758  ...  0.024242  0.078788  0.087879  0.021212   \n",
       "ABIZ01000001.1_6519  0.061453  ...   0.01676  0.097765  0.128492  0.036313   \n",
       "\n",
       "                         aa_E      aa_R      aa_Q      aa_N      aa_F  \\\n",
       "ABIZ01000001.1_1     0.070288  0.063898  0.044728  0.031949  0.031949   \n",
       "ABIZ01000001.1_2     0.017857  0.107143       NaN  0.035714  0.017857   \n",
       "ABIZ01000001.1_3     0.062914   0.05298  0.019868  0.039735  0.029801   \n",
       "ABIZ01000001.1_4     0.066852  0.064067  0.030641  0.050139  0.038997   \n",
       "ABIZ01000001.1_5     0.070796  0.084071  0.053097  0.017699  0.030973   \n",
       "...                       ...       ...       ...       ...       ...   \n",
       "ABIZ01000001.1_6515  0.021898  0.065693  0.029197  0.007299  0.058394   \n",
       "ABIZ01000001.1_6516       0.1  0.071429  0.071429  0.057143  0.028571   \n",
       "ABIZ01000001.1_6517  0.080882  0.095588  0.051471  0.014706  0.029412   \n",
       "ABIZ01000001.1_6518  0.075758  0.054545  0.018182  0.036364  0.036364   \n",
       "ABIZ01000001.1_6519  0.078212  0.039106  0.036313  0.039106  0.041899   \n",
       "\n",
       "                         aa_W  \n",
       "ABIZ01000001.1_1      0.01278  \n",
       "ABIZ01000001.1_2     0.035714  \n",
       "ABIZ01000001.1_3     0.013245  \n",
       "ABIZ01000001.1_4     0.011142  \n",
       "ABIZ01000001.1_5     0.022124  \n",
       "...                       ...  \n",
       "ABIZ01000001.1_6515  0.029197  \n",
       "ABIZ01000001.1_6516       NaN  \n",
       "ABIZ01000001.1_6517  0.022059  \n",
       "ABIZ01000001.1_6518  0.015152  \n",
       "ABIZ01000001.1_6519  0.011173  \n",
       "\n",
       "[6519 rows x 27 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_data = genome_calc.protein_data()\n",
    "df_proteins = pd.DataFrame(protein_data).T\n",
    "df_proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bd854707-2e95-4f65-8aeb-b2b18c33acd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAikAAAGfCAYAAACa6ELrAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8WgzjOAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAc3ElEQVR4nO3dbWxU553w4f8YakMb23UJNIlrQSKZbrakJn3/kBaaZFzVDitIVwnatLQglb5E3pa0XQGr7KaRGkIJpUo/tU0jtrsrEWlTt2ItBUdbGVWq+iK1o0JKZKMmFA1NCl6PJ4kcXOF5PvB0nsfYBPwCc499XdJ84Jx7xvcwMPy4z5kzmVKpVAoAgMTUVHoCAACTESkAQJJECgCQJJECACRJpAAASRIpAECSRAoAkKSqjZRSqRTFYjFc5gUA5qaqjZRXXnklGhsb45VXXqn0VACAK6BqIwUAmNtECgCQJJECACRJpAAASRIpAECSRAoAkCSRAgAkSaQAAEkSKQBAkkQKAJAkkQIAJEmkAABJEikAQJJECgCQJJECACRJpAAASRIpAECSFlZ6AsD8k8lkJmwrlUoVmAmQMispAECSRAoAkCSRAgAkSaQAAEkSKQBAkkQKAJAkH0EGrrjJPnIMcClWUgCAJIkUACBJDvcASXJVWsBKCgCQJJECACRJpAAASRIpAECSRAoAkCSRAgAkSaQAAEkSKQBAkkQKAJAkkQIAJEmkAABJEikAQJJECgCQJJECACRJpAAASRIpAECSRAoAkCSRAgAkaWGlJwAQEZHJZCo9BSAxM1pJOXv2bKxcuTKuueaacduPHDkS2Ww26uvrY+nSpbF58+YoFArjxhQKhdi8eXMsXbo06uvrI5vNxtGjR2cyHQBgDplRpOzevTsaGxvHbcvn87F27drIZrORz+cjl8vFyMhIdHZ2xtjYWEREjI2NRUdHR4yMjEQul4t8Ph/ZbDbWrFkTp06dmsmUAIA5IlMqlUrTueMf/vCHuPPOO+Pb3/52/MM//EO8+uqrERHR1dUVp0+fjgMHDpTHjo6ORmtra+zbty/uvvvuePrpp2Pbtm1x/PjxqK2tLY/buHFjLFu2LB5//PFL/vxisRiNjY0xPDwcDQ0N03kKwFUyW4dypvl2BVSpaa+kdHV1xUMPPTQhEA4ePBjr168ft622tjba29ujp6enPKajo2NcoERErFu3rjwGAJjfphUp3d3dMTQ0FJ/61KfGbR8dHY0TJ07ErbfeOuE+q1evjoGBgYiI6O/vv+iYF154Ic6dOzdh39mzZ6NYLI67AQBz15Qj5bXXXosHHnggvvOd70xYwh0aGoqImPTwS1NTU5w5c6Y87mJjSqVSDA4OTti3a9euaGxsLN9aWlqmOnUAoIpMOVIefvjhaG9vj/e+971XYj4XtWPHjhgeHi7fTp48eVV/PgBwdU3pOin9/f2xf//+eO655ybd39TUFBHnT2q9/vrrx+0rFApx7bXXlsdNdrimUChEJpOJJUuWTNhXV1cXdXV1U5kuAFDFphQpv//972NwcDBWrFhR3nbu3Ll4/fXX45prromNGzfG8uXLI5fLxTvf+c5x983lctHa2hoREStXroxcLjfh8XO5XNx4442xYMGCqT8TAGBOmVKkZLPZeP7558dt++Uvfxlbt26NXC4X9fX1sXjx4uju7o577723PGZ0dDQOHToU+/bti4jzn+LZtm1bjI6OjvuEz8GDB6Ozs3MmzwcAmCOmfZ2Uv+rr64u77rqrfJ2UfD4f7373u2P79u3xuc99Ll555ZX46le/Gn/84x/jZz/7WdTU1MTY2Fh8+MMfjpaWlti7d2/U19fHd7/73di9e3f87ne/ixtuuOGSP9d1UqB6uE4KMB2z/gWDzc3Ncfjw4ejt7Y3m5uZoa2uLRYsWRU9PT9TUnP9xNTU10dPTE4sXL462trZobm6O3t7e6Ovru6xAAQDmvhmvpFSKlRSoHlZSgOmY9ZUUAIDZIFIAgCSJFAAgSSIFAEiSSAEAkjSli7kBXOjCT+74BA4wW6ykAABJEikAQJJECgCQJJECACRJpAAASRIpAECSRAoAkCSRAgAkSaQAAEkSKQBAkkQKAJAkkQIAJMkXDAJV68IvN4zwBYcwl1hJAQCSJFIAgCSJFAAgSSIFAEiSSAEAkiRSAIAkiRQAIEkiBQBIkkgBAJLkirNA1ZjsCrPA3GUlBQBIkkgBAJIkUgCAJDknBZhVzhsBZouVFAAgSSIFAEiSSAEAkiRSAIAkiRQAIEkiBQBIkkgBAJIkUgCAJIkUACBJIgUASJJIAQCSJFIAgCSJFAAgSSIFAEiSSAEAkiRSAIAkiRQAIEkiBQBIkkgBAJIkUgCAJIkUACBJIgUASJJIAQCSJFIAgCSJFAAgSSIFAEiSSAEAkiRSAIAkiRQAIEkiBQBIkkgBAJIkUgCAJIkUACBJIgUASNKUI+Xs2bOxc+fOWL58eSxevDhWrVoV3//+98eNOXLkSGSz2aivr4+lS5fG5s2bo1AojBtTKBRi8+bNsXTp0qivr49sNhtHjx6d0ZMBAOaOKUfK9u3b45lnnonu7u4YGhqK733ve/Hwww/HE088ERER+Xw+1q5dG9lsNvL5fORyuRgZGYnOzs4YGxuLiIixsbHo6OiIkZGRyOVykc/nI5vNxpo1a+LUqVOz+wwBgKqUKZVKpanc4S9/+Uu89tpr8da3vrW87bHHHosf/ehH8fOf/zy6urri9OnTceDAgfL+0dHRaG1tjX379sXdd98dTz/9dGzbti2OHz8etbW15XEbN26MZcuWxeOPP37JeRSLxWhsbIzh4eFoaGiYylMAZlEmk6n0FMaZ4lsakLApr6S86U1vGhcoERHDw8Nx7bXXRkTEwYMHY/369eP219bWRnt7e/T09JTHdHR0jAuUiIh169aVx1zo7NmzUSwWx90AgLlrRifOvvrqq/GjH/0onnzyyfiXf/mXGB0djRMnTsStt946Yezq1atjYGAgIiL6+/svOuaFF16Ic+fOTdi3a9euaGxsLN9aWlpmMnVgGjKZzIQbwJUy7Uj58pe/HPX19XHPPffEI488Eu973/tiaGgoImLSwy9NTU1x5syZiIgYGhq66JhSqRSDg4MT9u3YsSOGh4fLt5MnT0536gBAFZh2pHz729+OwcHB+M///M/42te+Fk8++eRszmuCurq6aGhoGHcDAOauhTO589ve9ra49957Y9GiRbF169b45Cc/GRHnT2q9/vrrx40tFArl81aampomPaekUChEJpOJJUuWzGRaAMAcMCsXc3vve98bf/7zn2NoaCiWL18euVxuwphcLhetra0REbFy5cqLjrnxxhtjwYIFszEtAKCKTSlS/vSnP8WWLVvi5ZdfHrf9N7/5TbzlLW+JpqamWLduXXR3d4/bPzo6GocOHYrOzs6I+H+f4hkdHR037uDBg+UxAMD8NqVIWbZsWeRyubjvvvvi+eefj5GRkfjpT38a//iP/xgPPvhg1NbWxvbt2+PZZ5+NPXv2RLFYjHw+H5/+9KfjHe94R/mjyRs2bIiWlpbYtGlT5PP5KBaLsWfPnnj22Wdj+/btV+J5AgBVZkqRsmDBgvif//mfaG1tjTvuuCOWLFkSX/7yl2Pnzp3xT//0TxER0dzcHIcPH47e3t5obm6Otra2WLRoUfT09ERNzfkfV1NTEz09PbF48eJoa2uL5ubm6O3tjb6+vrjhhhtm/1kCAFVnylecTYUrzsLVVw3XRanStzRgEr4FGQBIkkgBAJIkUgCAJIkUACBJIgUASJJIAQCSJFIAgCSJFAAgSSIFAEiSSAEAkiRSAIAkiRQAIEkiBQBI0sJKTwBgNk32Tc2+GRmqk5UUACBJIgUASJJIAQCSJFIAgCSJFAAgSSIFAEiSSAEAkiRSAIAkiRQAIEkiBQBIkkgBAJIkUgCAJIkUACBJIgUASJJIAQCStLDSEwC40jKZzLhfl0qlCs0EmAorKQBAkkQKAJAkkQIAJEmkAABJEikAQJJECgCQJJECACRJpAAASRIpAECSRAoAkCSRAgAkSaQAAEkSKQBAkkQKAJAkkQIAJEmkAABJEikAQJJECgCQpIWVngCQrkwmU+kpAPOYlRQAIEkiBQBIkkgBAJLknBRg3pnsXJtSqVSBmQBvxEoKAJAkkQIAJEmkAABJEikAQJJECgCQJJECACRJpAAASRIpAECSRAoAkCSRAgAkSaQAAEkSKQBAkkQKAJCkaUXK/v3745Zbbok3v/nNcdNNN8WDDz4Yr7/+enn/kSNHIpvNRn19fSxdujQ2b94chUJh3GMUCoXYvHlzLF26NOrr6yObzcbRo0dn9GQAgLljypHyxBNPRFdXVzz88MNx+vTp+PGPfxzPPPNMfOELX4iIiHw+H2vXro1sNhv5fD5yuVyMjIxEZ2dnjI2NRUTE2NhYdHR0xMjISORyucjn85HNZmPNmjVx6tSp2X2GAEBVypRKpdJU7nDvvffGbbfdFl1dXeVtv/71r+NDH/pQDA0NxT//8z/H6dOn48CBA+X9o6Oj0draGvv27Yu77747nn766di2bVscP348amtry+M2btwYy5Yti8cff/yS8ygWi9HY2BjDw8PR0NAwlacAXKZMJlPpKVw1U3wrBK6CKa+kPPXUU+MCJSKioaGhvEpy8ODBWL9+/bj9tbW10d7eHj09PeUxHR0d4wIlImLdunXlMQDA/DYrJ87u378/2traYtGiRXHixIm49dZbJ4xZvXp1DAwMREREf3//Rce88MILce7cuQn7zp49G8VicdwNAJi7Zhwphw4dir1798Y3vvGNGBoaioiY9PBLU1NTnDlzJiIihoaGLjqmVCrF4ODghH27du2KxsbG8q2lpWWmUwcAEjajSDl8+HBs2LAhdu7cGZ2dnbM1p0nt2LEjhoeHy7eTJ09e0Z8HAFTWwune8Re/+EXcdddd8aUvfSkeeuihiDi/EhJx/qTW66+/ftz4QqEQ1157bXncZIdrCoVCZDKZWLJkyYR9dXV1UVdXN93pAgBVZlorKb/97W/j4x//eNx///2xa9eu8vba2tpYvnx55HK5CffJ5XLR2toaERErV6686Jgbb7wxFixYMJ1pAQBzyJQj5bnnnov29vbYunVrPProoxP2r1u3Lrq7u8dtGx0djUOHDpUPCf31Uzyjo6Pjxh08ePCKHzYCAKrDlK6Tks/n433ve1988pOfjD179lx0zLvf/e7Yvn17fO5zn4tXXnklvvrVr8Yf//jH+NnPfhY1NTUxNjYWH/7wh6OlpSX27t0b9fX18d3vfjd2794dv/vd7+KGG2645FxcJwWuPNdJASppSispAwMD8dJLL8Vjjz0WmUxmwq2vry+am5vj8OHD0dvbG83NzeWPJvf09ERNzfkfV1NTEz09PbF48eJoa2uL5ubm6O3tjb6+vssKFABg7pvyFWdTYSUFrjwrKUAl+RZkACBJIgUASJJIAQCSJFIAgCSJFAAgSSIFAEiSSAEAkiRSAIAkiRQAIEkiBQBIkkgBAJIkUgCAJIkUACBJIgUASJJIAQCSJFIAgCSJFAAgSSIFAEjSwkpPAEhDJpOp9BQq6sLnXyqVKjQT4K+spAAASRIpAECSRAoAkCSRAgAkSaQAAEkSKQBAkkQKAJAkkQIAJEmkAABJEikAQJJECgCQJJECACRJpAAASRIpAECSRAoAkCSRAgAkSaQAAElaWOkJAJWRyWQqPYWkTfb7UyqVKjATmL+spAAASRIpAECSRAoAkCSRAgAkSaQAAEkSKQBAkkQKAJAkkQIAJEmkAABJEikAQJJECgCQJN/dA3CZLvw+H9/lA1eWlRQAIEkiBQBIkkgBAJIkUgCAJIkUACBJIgUASJJIAQCS5DopMA9ceH0PgGpgJQUASJJIAQCSJFIAgCSJFAAgSSIFAEiSSAEAkiRSAIAkiRQAIEkiBQBI0rQjZXBwMNasWROrVq2asO/IkSORzWajvr4+li5dGps3b45CoTBuTKFQiM2bN8fSpUujvr4+stlsHD16dLrTAQDmmGlFyrFjx+KDH/xgnDhxYsK+fD4fa9eujWw2G/l8PnK5XIyMjERnZ2eMjY1FRMTY2Fh0dHTEyMhI5HK5yOfzkc1mY82aNXHq1KmZPSMAYE7IlEql0lTuUCwW46abbootW7bEzTffHHv37h23AtLV1RWnT5+OAwcOlLeNjo5Ga2tr7Nu3L+6+++54+umnY9u2bXH8+PGora0tj9u4cWMsW7YsHn/88cuaR2NjYwwPD0dDQ8NUngLMO76758qY4tsnMEVTXklpaGiIvr6++OY3vznpG9/Bgwdj/fr147bV1tZGe3t79PT0lMd0dHSMC5SIiHXr1pXHAADz27QO90x2HkrE+RWTEydOxK233jph3+rVq2NgYCAiIvr7+y865oUXXohz585N2Hf27NkoFovjbgDA3DWrn+4ZGhqKiJj08EtTU1OcOXOmPO5iY0qlUgwODk7Yt2vXrmhsbCzfWlpaZnPqAEBiquYjyDt27Ijh4eHy7eTJk5WeEgBwBS2czQdramqKiPMntV5//fXj9hUKhbj22mvL4yY7XFMoFCKTycSSJUsm7Kurq4u6urrZnC4AkLBZXUmpra2N5cuXRy6Xm7Avl8tFa2trRESsXLnyomNuvPHGWLBgwWxOCwCoQrN+uGfdunXR3d09btvo6GgcOnQoOjs7y2N6enpidHR03LiDBw+WxwAA89usR8r27dvj2WefjT179kSxWIx8Ph+f/vSn4x3veEf5o8kbNmyIlpaW2LRpU+Tz+SgWi7Fnz5549tlnY/v27bM9JQCgCs16pDQ3N8fhw4ejt7c3mpubo62tLRYtWhQ9PT1RU3P+x9XU1ERPT08sXrw42traorm5OXp7e6Ovry9uuOGG2Z4SAFCFpnzF2VS44ixcPlecvTKq9O0TqkbVfAQZAJhfRAoAkCSRAgAkSaQAAEkSKQBAkkQKAJCkWf3uHoD5ZLKPdvtYMsweKykAQJJECgCQJJECACRJpAAASXLiLMxBvqunci78vXciLUyflRQAIEkiBQBIkkgBAJIkUgCAJIkUACBJIgUASJJIAQCSJFIAgCSJFAAgSSIFAEiSSAEAkuS7ewCuoMm+R8n3+cDlsZICACRJpAAASXK4BxJ2OYcKJhsDMBdYSQEAkiRSAIAkiRQAIEkiBQBIkhNnAa6yC092dt0UmJxIAagwF3yDyYkUgHnOyg6pck4KAJAkkQIAJEmkAABJck4KVBmXwZ8fnCcCVlIAgESJFAAgSQ73AMxhDhtRzaykAABJEikAQJIc7gGoAlfz0vku008qrKQAAEkSKQBAkhzugQqxpA7wxkQKwBzhasTMNSIFgEu6nACyEshsEykAzAoXjmO2OXEWAEiSlRQArggnhzNTIgUS4sRHpsKfF+Y6h3sAgCRZSQGYR6y+UE2spAAASbKSAsBV42PKTIWVFAAgSVZS4Apw3B9g5kQKAElxSIi/EikwRS5QBbPnclYd/Z2bv5yTAgAkyUoKzALnoADMPispAECSRAoAkCSHe5g3LufkOyfoQXXwCaD5oaIrKd3d3bF69epYvHhxrFixIh555BF/0ACYskwmM+FG9avYSkpPT09s2bIl/uM//iM++tGPxsDAQGzatCmKxWI8+uijlZoWVWK2/hc13Y8/AnODFZm0ZUoVekVuueWWuP/+++Pzn/98eduxY8eira0tTp48GW9/+9vf8P7FYjEaGxtjeHg4GhoarvR0uYqmEwWT/TEWF8BUiZS0VORwz4svvhhHjx6N9evXj9t+8803x/Lly6O3t7cS0xrH0uHV4fcZSIn3pLRU5HBPf39/XHfddXHddddN2Ld69eoYGBiYsP3s2bNx9uzZ8q+Hh4cj4vyKytVyNX/W1dTY2Dju13/9vX2jMZO58H6Xc5/Z4o0EuFIu571/Ou+Rl/M407nP5bqc9+zL+fnTVV9ff8n37opEytDQ0EUP0TQ1NcWZM2cmbN+1a1d8/etfn7C9paVl1ud3MVfzH91Kmu7znC+/P8D8MlvvbdN5nCv5vno5j30lf/7lnK5RNR9B3rFjRzzwwAPlX4+NjcX//u//xpIlS+bc/6KLxWK0tLTEyZMnnW9TBbxe1cXrVV28XtVjqq9VfX39JcdUJFKampouunxWKBTib/7mbyZsr6uri7q6unHb3vrWt16J6SWjoaHBX8oq4vWqLl6v6uL1qh6z+VpV5MTZlStXxksvvRQvv/zyhH25XC5aW1srMCsAICUViZQVK1bEqlWroru7e9z2Y8eOxYsvvhjt7e2VmBYAkJCKnZOye/fuuO+++6KlpSVuv/326O/vj02bNsVXvvKVS14jZa6rq6uLf/3Xf51weIs0eb2qi9eruni9qseVeK0qdjG3iPOXxX/ooYeiv78/li1bFlu3bo2dO3fOuRNhAYCpq2ikAABcTEW/YBAA4GJECgCQJJECACRJpCTuv//7v+MDH/hAXHPNNdHS0hJdXV0xNDRU6WnxBkZGRuKee+6Ja665ptJT4QLd3d2xevXqWLx4caxYsSIeeeQR33qbuMHBwVizZk2sWrWq0lPhEvbv3x+33HJLvPnNb46bbropHnzwwXj99ddn9JgiJWGHDh2Kv//7v4/Pfvaz8ac//Sn6+vri+PHj8YlPfKLSU+MiTp06FR/5yEfiV7/6VaWnwgV6enpiy5Yt8Y1vfCMGBwfjJz/5STz11FOxY8eOSk+Nizh27Fh88IMfjBMnTlR6KlzCE088EV1dXfHwww/H6dOn48c//nE888wz8YUvfGFGj+vTPQl74IEHora2Nh599NHytj//+c/x9re/PY4cOeJ/Fok5d+5crFq1Kt7znvfEZz7zmdiwYUO8+uqrlZ4W/9ctt9wS999/f3z+858vbzt27Fi0tbXFyZMn5/31mVJTLBbjpptuii1btsTNN98ce/fujaNHj1Z6WlzEvffeG7fddlt0dXWVt/3617+OD33oQ2/4pcKXUjVfMDgffetb35qwrb6+PmpqaixRJ2jBggXxX//1X/G3f/u3cfjw4UpPh//Piy++GEePHo3169eP237zzTfH8uXLo7e3Nz71qU9VZnJMqqGhIfr6+mLVqlWxf//+Sk+HS3jqqacmbGtoaIixsbEZPa5IqTI//OEPY8mSJZN+CSOV9653vavSU2AS/f39cd1118V11103Yd/q1atjYGCgArPiUqwWV7f9+/dHW1vbjL5sUKRUkVwuF1/72tfi0UcfjTe96U2Vng5UjTdabm5qaoozZ85c5RnB3Hbo0KHYu3fvhO/omyonzlbAnXfeGZlM5qK3tWvXTrjP73//+8hms/GJT3wivvjFL179Sc9j03m9AOarw4cPx4YNG2Lnzp3R2dk5o8eyklIBvb29b3ic7sLvLjp+/Hjceeed8bGPfSx+8IMfXOnpcYGpvl6kp6mpKYrF4qT7CoWCw6cwS37xi1/EXXfdFV/60pfioYcemvHjWUmpgJqamli4cOFFbwsWLCiPPXHiRNx+++3x0Y9+NP7t3/4tamq8ZFfbVF4v0rRy5cp46aWX4uWXX56wL5fLRWtrawVmBXPLb3/72/j4xz8e999/f+zatWtWHtO/eAk7depU3HHHHXHbbbfFD3/4Q/8YwjStWLEiVq1aNeH4+LFjx+LFF1+M9vb2Cs0M5obnnnsu2tvbY+vWreMumzFTDvck6rXXXos77rgj3v/+98e///u/CxSYod27d8d9990XLS0tcfvtt0d/f39s2rQpvvKVr7hGCsxAPp+PO++8Mz7zmc/E7t27Z/WxraQk6vTp0/H888/HgQMHYuHChRNO1nTdAJiajo6OePLJJ2Pnzp3xtre9Lf7u7/4u7rnnnnjkkUcqPTWoagMDA/HSSy/FY489NumHC/r6+qb92K44CwAkyUoKAJAkkQIAJEmkAABJEikAQJJECgCQJJECACRJpAAASRIpAECSRAoAkCSRAgAkSaQAAEkSKQBAkv4Pqma1N4iOch8AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "x = 'gravy'\n",
    "\n",
    "X = df_proteins[x].values \n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "ax.hist(X, bins=100)\n",
    "#ax.axvline(0.5)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e3bc8d6-44e8-49ab-a470-f33b05308143",
   "metadata": {},
   "source": [
    "# Datasets for Training and Testing\n",
    "\n",
    "## Balancing a Dataset by Taxonomy\n",
    "\n",
    "Removes genomes from taxonomic lineages overrepresented in the dataset relative to Genome Taxonomy Database representative species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a398ba00-3ce0-4bc4-bf0a-ddf84c7d2ba8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GCA_000979555': ('Archaea',\n",
       "  'Halobacteriota',\n",
       "  'Methanosarcinia',\n",
       "  'Methanosarcinales',\n",
       "  'Methanosarcinaceae',\n",
       "  'Methanosarcina',\n",
       "  'Methanosarcina mazei'),\n",
       " 'GCA_001315865': ('Archaea',\n",
       "  'Halobacteriota',\n",
       "  'Methanosarcinia',\n",
       "  'Methanosarcinales',\n",
       "  'Methanosarcinaceae',\n",
       "  'Methanosarcina',\n",
       "  'Methanosarcina mazei')}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Download taxonomy files\n",
    "from genome_spot.taxonomy.taxonomy import TaxonomyGTDB\n",
    "taxonomy = TaxonomyGTDB()\n",
    "\n",
    "{k:v for n, (k, v) in enumerate(taxonomy.taxonomy_dict.items()) if n < 2}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a0a326ac-0aed-49b3-816b-25b6ed868915",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.5"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "from genome_spot.taxonomy.balance import BalanceTaxa\n",
    "\n",
    "with open(\"../tests/test_data/test_genome_accessions.txt\", 'r') as fh:\n",
    "    genomes = [line.strip() for line in fh.readlines()]\n",
    "\n",
    "taxonomy = TaxonomyGTDB()\n",
    "balancer = BalanceTaxa(taxonomy=taxonomy)\n",
    "balanced_genomes = balancer.balance_dataset(\n",
    "    genomes=genomes,\n",
    "    proportion_to_keep=0.5,\n",
    "    diversity_rank=\"species\"\n",
    ")\n",
    "len(balanced_genomes) / len(genomes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d532f1fe-9866-4653-aafd-0f5167944639",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# It's reproducibly random\n",
    "\n",
    "balanced_genomes1 = BalanceTaxa(taxonomy=taxonomy).balance_dataset(\n",
    "    genomes=genomes,\n",
    "    proportion_to_keep=0.5,\n",
    "    diversity_rank=\"species\"\n",
    ")\n",
    "\n",
    "balanced_genomes2 = BalanceTaxa(taxonomy=taxonomy).balance_dataset(\n",
    "    genomes=genomes,\n",
    "    proportion_to_keep=0.5,\n",
    "    diversity_rank=\"species\"\n",
    ")\n",
    "\n",
    "set(balanced_genomes1).difference(balanced_genomes2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5e96cd9-c99a-4bf6-9e08-8f420fd335f5",
   "metadata": {},
   "source": [
    "## Test-Train Partition (Split) by Taxonomy\n",
    "\n",
    "Use balanced genomes\n",
    "\n",
    "NOTE: The script `make_holdout_sets.py` uses information from the variables themselves to decide which genomes to include and exclude for each condition. For simplicity, here only phylogeny is used and the growth condition is not considered."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "bf7d64b1-e769-490c-aa0a-7cbd9a25870d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.2 0.8\n",
      "20 30920\n"
     ]
    }
   ],
   "source": [
    "from genome_spot.taxonomy.partition import PartitionTaxa\n",
    "\n",
    "test_set = partitioner.partition(balanced_genomes, partition_size=0.2)\n",
    "train_set = sorted(set(balanced_genomes).difference(test_set))\n",
    "\n",
    "print(len(test_set) / len(balanced_genomes), len(train_set) / len(balanced_genomes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ac42ea2f-ded5-4a90-8217-011a0db576e3",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'partitioner' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# It's reproducibly random\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m partitioned_genomes1 \u001b[38;5;241m=\u001b[39m \u001b[43mpartitioner\u001b[49m\u001b[38;5;241m.\u001b[39mpartition(balanced_genomes, partition_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.2\u001b[39m)\n\u001b[1;32m      3\u001b[0m partitioned_genomes2 \u001b[38;5;241m=\u001b[39m partitioner\u001b[38;5;241m.\u001b[39mpartition(balanced_genomes, partition_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.2\u001b[39m)\n\u001b[1;32m      4\u001b[0m partitioned_genomes1 \u001b[38;5;241m==\u001b[39m partitioned_genomes2\n",
      "\u001b[0;31mNameError\u001b[0m: name 'partitioner' is not defined"
     ]
    }
   ],
   "source": [
    "# It's reproducibly random\n",
    "partitioned_genomes1 = partitioner.partition(balanced_genomes, partition_size=0.2)\n",
    "partitioned_genomes2 = partitioner.partition(balanced_genomes, partition_size=0.2)\n",
    "partitioned_genomes1 == partitioned_genomes2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "371a6b42-d75f-4c53-8e09-42f34b799ae0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Some helpers to look up stuff\n",
    "extended_partitioned_genomes = sorted(partitioner.find_relatives_of_partitioned_set_in_reference(test_set))\n",
    "print(len(test_set), len(extended_partitioned_genomes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "fbe27f78-ca9b-485b-bf1b-f3715a6e5319",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Good!\n"
     ]
    }
   ],
   "source": [
    "# check for data leakage\n",
    "partitioned_taxa = set(taxonomy.taxa_of_genomes(partitioned_genomes, partitioner.partition_rank))\n",
    "nonpartitioned_taxa = set(taxonomy.taxa_of_genomes(nonpartitioned_genomes, partitioner.partition_rank))\n",
    "assert len(partitioned_taxa.intersection(nonpartitioned_taxa)) == 0\n",
    "\n",
    "# check partitioned genomes all present\n",
    "len(set(partitioned_genomes).intersection(set(extended_partitioned_genomes))) == len(set(partitioned_genomes))\n",
    "assert len(set(partitioned_genomes).difference(set(extended_partitioned_genomes))) == 0\n",
    "print('Good!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dcb2df37-301f-41e1-88b2-0689cf778e21",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'balancer' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m proportion_dict \u001b[38;5;241m=\u001b[39m \u001b[43mbalancer\u001b[49m\u001b[38;5;241m.\u001b[39massess_proportion(subset_genomes\u001b[38;5;241m=\u001b[39mtest_set, reference_genomes\u001b[38;5;241m=\u001b[39mtest_set\u001b[38;5;241m.\u001b[39munion(train_set), rank\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mphylum\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      2\u001b[0m test_diversity \u001b[38;5;241m=\u001b[39m taxonomy\u001b[38;5;241m.\u001b[39mmeasure_diversity(query_rank\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfamily\u001b[39m\u001b[38;5;124m'\u001b[39m, diversity_rank\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mspecies\u001b[39m\u001b[38;5;124m'\u001b[39m, subset_genomes\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mlist\u001b[39m(test_set))\n\u001b[1;32m      3\u001b[0m train_diversity \u001b[38;5;241m=\u001b[39m taxonomy\u001b[38;5;241m.\u001b[39mmeasure_diversity(query_rank\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfamily\u001b[39m\u001b[38;5;124m'\u001b[39m, diversity_rank\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mspecies\u001b[39m\u001b[38;5;124m'\u001b[39m, subset_genomes\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mlist\u001b[39m(train_set))\n",
      "\u001b[0;31mNameError\u001b[0m: name 'balancer' is not defined"
     ]
    }
   ],
   "source": [
    "proportion_dict = balancer.assess_proportion(subset_genomes=test_set, reference_genomes=test_set.union(train_set), rank='phylum')\n",
    "test_diversity = taxonomy.measure_diversity(query_rank='family', diversity_rank='species', subset_genomes=list(test_set))\n",
    "train_diversity = taxonomy.measure_diversity(query_rank='family', diversity_rank='species', subset_genomes=list(train_set))\n",
    "\n",
    "fig, axes = plt.subplots(1,2, figsize=(12,3),gridspec_kw={'width_ratios' : (3,1)})\n",
    "test_color = 'tomato'\n",
    "train_color = 'skyblue'\n",
    "ax = axes[0]\n",
    "df_counts = pd.DataFrame.from_dict(proportion_dict, orient='index').sort_index()\n",
    "df_counts.columns = ['Test']\n",
    "df_counts['Train'] = 1 - df_counts['Test']\n",
    "ax.set_title(f\"Distribution of Species by Partition across Phyla\", fontsize=10)\n",
    "df_counts.plot.bar(ax=ax, width=1, color=[test_color, train_color], edgecolor='None', stacked=True, legend=False)\n",
    "ax.set_ylim([0,1])\n",
    "ax.set_xlim([-0.5,ax.get_xlim()[1]-0.45])\n",
    "\n",
    "ax = axes[1]\n",
    "X1 = [0] + list(pd.DataFrame.from_dict(test_diversity, orient='index').sort_values(by=0, ascending=False)[0].values)\n",
    "X2 = [0] + list(pd.DataFrame.from_dict(train_diversity, orient='index').sort_values(by=0, ascending=False)[0].values)\n",
    "ax.set_title(f\"Distribution of Species across Families\", fontsize=10)\n",
    "ax.plot(range(len(X1)), np.cumsum(X1) / np.sum(X1), marker='.', color=test_color, label='Test')\n",
    "ax.plot(range(len(X2)), np.cumsum(X2) / np.sum(X2), marker='.', color=train_color, label='Train')\n",
    "ax.set_ylabel('Cumulative Genomes')\n",
    "ax.set_xlabel('Family')\n",
    "ax.set_ylim([0,ax.get_ylim()[1]])\n",
    "ax.set_xlim([0,ax.get_xlim()[1]])\n",
    "\n",
    "for ax in axes:\n",
    "    ax.legend(frameon=True)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    \n",
    "plt.show() "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "432c326d-580d-44cb-82ba-ec5cd250f586",
   "metadata": {},
   "source": [
    "## Creating Cross-Validation Sets by Taxonomy\n",
    "\n",
    "The test set is your final **final** look at the model quality. You should use cross-valdation (look it up) to evaluate and improve models. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b7adbc67-b1ea-46ee-a88a-4764b4060816",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fold 1\n",
      "64 genomes in training\n",
      "16 genomes in validation\n",
      "0.2 fraction held out\n",
      "\n",
      "Fold 2\n",
      "62 genomes in training\n",
      "18 genomes in validation\n",
      "0.225 fraction held out\n",
      "\n",
      "Fold 3\n",
      "64 genomes in training\n",
      "16 genomes in validation\n",
      "0.2 fraction held out\n",
      "\n",
      "Fold 4\n",
      "65 genomes in training\n",
      "15 genomes in validation\n",
      "0.1875 fraction held out\n",
      "\n",
      "Fold 5\n",
      "65 genomes in training\n",
      "15 genomes in validation\n",
      "0.1875 fraction held out\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# As run in this function:\n",
    "# from genome_spot.model_training.make_holdout_sets import make_cv_sets_by_phylogeny\n",
    "# cv_sets = make_cv_sets_by_phylogeny(genomes=np.array(list(balanced_genomes)), partition_rank='order', kfold = 10)\n",
    "\n",
    "genomes = train_set\n",
    "kfold = 5\n",
    "partitioner = PartitionTaxa(\n",
    "        taxonomy=TaxonomyGTDB(),\n",
    "        partition_rank='family',\n",
    "        diversity_rank=\"species\",\n",
    "    )\n",
    "\n",
    "cv_sets = []  # training, validation\n",
    "remaining_genomes = set(genomes)\n",
    "fold_sizes = []\n",
    "for k in range(kfold):\n",
    "    validation_partition_size = 1 / (kfold - k)\n",
    "    partitioned_genomes = partitioner.partition(remaining_genomes, partition_size=validation_partition_size)\n",
    "    if k < kfold - 1:\n",
    "        remaining_genomes = set(remaining_genomes).difference(partitioned_genomes)\n",
    "    else:\n",
    "        partitioned_genomes = remaining_genomes\n",
    "    validation_indices = np.in1d(genomes, np.array(list(partitioned_genomes))).nonzero()[0]\n",
    "    training_indices = np.in1d(genomes, np.array(list(set(genomes).difference(partitioned_genomes)))).nonzero()[0]\n",
    "    cv_sets.append((training_indices, validation_indices))\n",
    "    fold_sizes.append(len(partitioned_genomes) / len(genomes))\n",
    "\n",
    "    print(f\"Fold {k+1}\")\n",
    "    print(f\"{len(training_indices)} genomes in training\")\n",
    "    print(f\"{len(validation_indices)} genomes in validation\")\n",
    "    print(f\"{len(partitioned_genomes) / len(genomes)} fraction held out\")\n",
    "    print()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea301147-e470-4f9f-8389-c0aeb0cf53a3",
   "metadata": {},
   "source": [
    "## BacDive Trait Data\n",
    "### Download Trait Data from BacDive API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f0b88f25-bcb9-4208-9b86-203636dddd76",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-03-15 15:39:45 INFO Logging into BacdiveClient\n",
      "2024-03-15 15:39:47 INFO Iniating 100 queries in 1 chunks\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-- Authentication successful --\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-03-15 15:39:50 INFO Searching query indices 0-100 returned 100 results\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "dict_keys(['General', 'Name and taxonomic classification', 'Morphology', 'Culture and growth conditions', 'Physiology and metabolism', 'Isolation, sampling and environmental information', 'Safety information', 'Sequence information', 'Genome-based predictions', 'External links', 'Reference'])"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from genome_spot.model_training.download_trait_data import QueryBacDive\n",
    "\n",
    "credentials_filepath = f'../../.bacdive_credentials'\n",
    "\n",
    "with open(credentials_filepath) as fh:\n",
    "    username = fh.readline().strip()\n",
    "    password = fh.readline().strip()\n",
    "\n",
    "\n",
    "min_bacdive_id = 100\n",
    "max_bacdive_id = 200\n",
    "bacdive_dict = QueryBacDive(\n",
    "            username=username, password=password,\n",
    "            max_bacdive_id=int(max_bacdive_id),\n",
    "            min_bacdive_id=int(min_bacdive_id),\n",
    "        ).scrape_bacdive_api()\n",
    "\n",
    "strain_ids = list(bacdive_dict.keys())\n",
    "strain_data = bacdive_dict[strain_ids[0]]\n",
    "strain_data.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d59efe2a-9119-4d33-b8f1-2a5a09fd780e",
   "metadata": {},
   "source": [
    "### Parse Trait Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6ba1f0e1-b5b2-46f8-be26-8ebbdca4b051",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ncbi_accession': 'GCA_900475915.1',\n",
       " 'ncbi_taxid': 644284,\n",
       " 'strain_id': 197,\n",
       " 'species': 'Arcanobacterium haemolyticum',\n",
       " 'ph_optimum': None,\n",
       " 'ph_optimum_min': None,\n",
       " 'ph_optimum_max': None,\n",
       " 'temperature_optimum': 37.0,\n",
       " 'salinity_optimum': None,\n",
       " 'salinity_midpoint': None,\n",
       " 'salinity_min': None,\n",
       " 'salinity_max': None,\n",
       " 'ph_min': None,\n",
       " 'ph_max': None,\n",
       " 'temperature_min': 37.0,\n",
       " 'temperature_max': 37.0,\n",
       " 'oxygen': 1,\n",
       " 'use_ph': False,\n",
       " 'use_temperature': False,\n",
       " 'use_salinity': False,\n",
       " 'use_oxygen': True,\n",
       " 'aerobe': None,\n",
       " 'anaerobe': 1,\n",
       " 'microaerophile': None,\n",
       " 'facultative anaerobe': 1,\n",
       " 'obligate aerobe': None,\n",
       " 'obligate anaerobe': None,\n",
       " 'facultative aerobe': None,\n",
       " 'aerotolerant': None,\n",
       " 'microaerotolerant': None}"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from genome_spot.model_training.download_trait_data import ComputeBacDiveTraits\n",
    "\n",
    "strain_data = bacdive_dict[strain_ids[2]]\n",
    "strain_traits = ComputeBacDiveTraits(strain_data).compute_trait_data()\n",
    "strain_traits"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv_physicochemistry",
   "language": "python",
   "name": "venv_physicochemistry"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
